mw_basic_auth.go

```diff
@@ -117,6 +111,66 @@ func (k *BasicAuthKeyIsValid) ProcessRequest(w http.ResponseWriter, r *http.Requ
 		ctxSetAuthToken(r, keyName)
 	}
 
-	// Request is valid, carry on
 	return nil, http.StatusOK
 }
+
+func (k *BasicAuthKeyIsValid) handleAuthFail(w http.ResponseWriter, r *http.Request, token string) (error, int) {
+
+	// Fire Authfailed Event
+	AuthFailed(k, r, token)
+
+	// Report in health check
+	reportHealthValue(k.Spec, KeyFailure, "-1")
+
+	return k.requestForBasicAuth(w, "User not authorised")
+}
+
+func (k *BasicAuthKeyIsValid) compareHashAndPassword(hash string, password string, logEntry *logrus.Entry) error {
+
+	cacheEnabled := !k.Spec.BasicAuth.DisableCaching
+	cacheDuration := time.Duration(60) * time.Second
+	if k.Spec.BasicAuth.CacheTTL > 0 {
+		cacheDuration = time.Duration(k.Spec.BasicAuth.CacheTTL) * time.Second
+	}
+
+	hasher := murmur3.New32()
+	passwordBytes := []byte(password)
+
+	doBcrypt := func(cacheEnabled bool, cacheDuration time.Duration, hashedPassword string, password []byte) error {
```

Comment:
what the purpose of this function literal? the reason I am asking - I don't see it passed anywhere as param or any closure to outer-scoped variables

Reply:
we need to call `doBcrypt` 3 times in the function. If you look at previous way was implemented, all of content from the function literal was being repeated many times. I didn't want to create pointer receiver function on the struct, because this is the only place that it is used.

I didn't want to place as separate pointer receiver, as is not used elsewhere.

Also, it makes the business logic bit of the code much easier to read & understand.